src/Sarif.ValidationTests/DefaultObjectPopulatingVisitor.cs

```diff
@@ -0,0 +1,293 @@
+
+            string jsonTypeName = GetJsonNameFor(objectTypeName);
+            string jsonPropertyName = GetJsonNameFor(propertyName);
+
+            // If _schema.Definitions does not contain the jsonTypeName, we are operating
+            // against the root sarifLog schema, which is what's stored in _schema
+            JsonSchema propertySchema = _schema.Definitions.ContainsKey(jsonTypeName) ? _schema.Definitions[jsonTypeName] : _schema;
+
+            return propertySchema.Required != null && propertySchema.Required.Contains(jsonPropertyName);
+        }
+
+        private string GetJsonNameFor(string name)
+        {
+            return name.Substring(0, 1).ToLower() + name.Substring(1);
+        }
+
+        private void AddElementToDictionary(object dictionary, object dictionaryValue)
+        {
+            var dictionaryType = dictionary.GetType();
+            MethodInfo method = dictionaryType.GetMethod("Add");
+            method.Invoke(dictionary, new[] { "PlaceholderDictionaryKey", dictionaryValue });
+        }
+
+        private void AddElementToList(object list, object listElement)
+        {
+            var listType = list.GetType();
+            MethodInfo method = listType.GetMethod("Add");
+            method.Invoke(list, new [] { listElement});
+        }
+
+        private object CreateEmptyList(Type propertyType)
+        {
+            // The SARIF OM defines various properties as IList<T>. This abstract type
+            // cannot be instantiated directly. We will construct an instance of the 
+            // concrete List<T> class, using the generic type argument from the OM.
+            // These constructed types are creatable via a parameterless constructor
+            // (which is what is ultimately invoked by Activator.CreateInstance).
+            var listType = typeof(List<>);
+            var constructedType = listType.MakeGenericType(propertyType.GenericTypeArguments[0]);
+            return Activator.CreateInstance(constructedType);
+        }
+
+        private object CreateEmptyDictionary(Type propertyType)
+        {
+            // The SARIF OM defines various properties as IDictionary<string, T>. This abstract
+            // type cannot be instantiated directly. We will construct an instance of the 
+            // concrete Dictionary<string, T> class, using the generic type argument from the OM.
+            // These constructed types are creatable via a parameterless constructor
+            // (which is what is ultimately invoked by Activator.CreateInstance).
+            var listType = typeof(Dictionary<,>);
+            var constructedType = listType.MakeGenericType(propertyType.GenericTypeArguments[0], propertyType.GenericTypeArguments[1]);
+            return Activator.CreateInstance(constructedType);
+        }
+
+        private static bool IsList(Type propertyType)
+        {
+            return propertyType.Name.StartsWith("IList");
```

Comment:
>StartsWith [](start = 37, length = 10)

Isn't there anything in the `Type` object that gives you "`IList`" exactly? This will match "`IListicle`". Similarly for `IDictionary`.

Reply:
Well, sure, now that you mention it. :)  I

            return propertyType.IsGenericType && propertyType.GetGenericTypeDefinition() == typeof(IList<>);


---
In reply to: [229132248](https://github.com/Microsoft/sarif-sdk/pull/1121#discussion_r229132248) [](ancestors = 229132248)